logstash-core/src/main/java/org/logstash/plugins/factory/RubyCodecDelegator.java

```diff
@@ -0,0 +1,137 @@
+        // invoke Ruby's codec #decode(data, block) and use a Block to capture the yielded LogStash::Event to
+        // back to Java and pass to the eventConsumer.
+        if (buffer.remaining() == 0) {
+            // no data to decode
+            return;
+        }
+
+        // setup the block callback bridge to invoke eventConsumer
+        final Block consumerWrapper = new Block(new JavaInternalBlockBody(currentContext.runtime, Signature.ONE_ARGUMENT) {
+            @Override
+            @SuppressWarnings("unchecked")
+            public IRubyObject yield(ThreadContext context, IRubyObject[] args) {
+                // Expect only one argument, the LogStash::Event instantiated by the Ruby codec
+                final IRubyObject event = args[0];
+                // TODO should we check the type of the args[0] or assume it's always an RubyEvent?
+                final Map<String, Object> unwrappedEvent = ((JrubyEventExtLibrary.RubyEvent) event).ruby_to_hash(currentContext)
+                        .toJava(Map.class);
+                eventConsumer.accept(unwrappedEvent);
+                return event;
+            }
+        });
+
+        byte[] byteInput = new byte[buffer.remaining()];
+        buffer.get(byteInput);
+        final RubyString data = RubyUtil.RUBY.newString(new String(byteInput));
+        // this causes problem in passing data, it trigger an invocation to #force_encoding if the data is not
+        // already converted to RubyString before passing it
+//                final IRubyObject data = JavaUtil.convertJavaToRuby(RubyUtil.RUBY, byteInput);
+        IRubyObject[] methodParams = new IRubyObject[]{data};
+        pluginInstance.callMethod(this.currentContext, "decode", methodParams, consumerWrapper);
+    }
+
+    @Override
+    public void flush(ByteBuffer buffer, Consumer<Map<String, Object>> eventConsumer) {
+        decode(buffer, eventConsumer);
+    }
+
+    @Override
+    @SuppressWarnings({"uncheked", "rawtypes"})
+    public void encode(Event event, OutputStream output) throws IOException {
+        // convert co.elastic.logstash.api.Event to JrubyEventExtLibrary.RubyEvent
+        if (!(event instanceof org.logstash.Event)) {
+            throw new IllegalStateException("The object to encode must be of type org.logstash.Event");
+        }
+
+        final JrubyEventExtLibrary.RubyEvent rubyEvent = JrubyEventExtLibrary.RubyEvent.newRubyEvent(currentContext.runtime, (org.logstash.Event) event);
+        final RubyArray param = RubyArray.newArray(currentContext.runtime, rubyEvent);
+        final RubyArray encoded = (RubyArray) pluginInstance.callMethod(this.currentContext, "multi_encode", param);
+
+        // method return an nested array, the outer contains just one element
+        // while the inner contains the original event and encoded event in form of String
+        final String result = (String) ((RubyArray) encoded.get(0)).get(1);
+        output.write(result.getBytes(StandardCharsets.UTF_8));
```

Comment:
avoids intermediate to Java string conversion + an extra `char[]` -> `byte[]` copy:
```suggestion
        final RubyString result = ((RubyArray) encoded.eltInternal(0)).eltInternal(1).convertToString();
        output.write(result.getByteList().getUnsafeBytes(), result.getByteList().getBegin(), result.getByteList().getRealSize());
```